### DISCUSSION TITLE
Preview HTML in PHP files

### DISCUSSION THREAD

**Discussion by nbernaz** [phcode-dev/phoenix#1877] [Ideas]
Would it be possible to allow the preview of a PHP file that contains valid HTML too?
I usually have a mixture of PHP and HTML in my files, and I don't care so much to be able to run the PHP but I would like to be able to render the HTML - and ignore any of the PHP code. Right now, if I rename the PHP file .html, some of the PHP is interpreted as HTML, and if I keep the PHP extension, the preview altogether does not work.

**Comment by abose** [comment]
Removing the php sections may be a bit tricky to do. but it is possible to enable live preview of php files as HTML, but that will display all the php text in the live preview as you mentioned.

**Comment by nbernaz** [comment]
Thank you. Maybe this is not the way to do things, but all my PHP are embedded in a single large <?php></php> section at the beginning of the file for readability, and I tought that you could add a rule just to ignore that. I understand that others may sprinkle php code here and there and that it would be a headache to remove, but I thought a single section would be easy.
The workaround that I found is to remove all the php to have proper rendering and then add it later at production time which is not headache because all the php is in one section, but still a pain to do.
Other than that, I just love your editor - it made my life so easy, considering that I was using notepad before.

**Comment by abose** [comment]
Thanks for checking out Phoenix Code and glad that it has made life Easier. 

Handling single PHP blocks in the way described isn't as straightforward as it may seem.

The issue with adding a special rule to ignore a single large <?php> block is that if we create an exception for one case, it sets an expectation that all PHP code, no matter how it's structured, would be treated similarly. From the editor's perspective, we need to ensure consistent and predictable behavior in every situation. If we introduced a feature that only works in some cases (like with a single PHP block), it could lead to confusion when things break. Users might not be sure if the issue lies in their code or if the editor is handling the PHP incorrectly.

One of our core goals with Phoenix Code is to provide rock-solid reliability so you can trust that when something isn’t working, it's likely due to your code, not the editor. Implementing partial solutions or hacks could compromise that trust and lead to a poor user experience.

[We have a tacking feature request to enable advanced PHP code intelligence and live previews in the future](https://github.com/phcode-dev/phoenix/issues/1768). So all these are likely to be addressed when we start on that. That may not come in the next 4 months though. 

If you have any other suggestions or encounter any issues, please don't hesitate to reach out.

**Comment by Olaes68** [comment]
Hi there,
I'm new on this. working on a project and I'm getting this error:                                                                                                                                  **failed to tokenize "<!DOCTYPE ...", state TEXT failed to consume data or change state. (parser-error)**
any idea what the issue is?

Thanks for your help.

Here is the code the issue is in line 14.

<?php
include 'db.php';

if ($_SERVER['REQUEST_METHOD'] === 'POST') {
    $name = $_POST['name'];
    $query = "INSERT INTO Equipment (name) VALUES (:name)";
    $stmt = $conn->prepare($query);
    $stmt->execute(['name' => $name]);
    header("Location: index.php");
    exit();
}
?>

(here) <!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link rel="stylesheet" href="styles.css">
    <title>Add Equipment</title>
</head>
<body>
    <h1>Add Equipment</h1>
    <form method="POST">
        <label for="name">Equipment Name:</label>
        <input type="text" id="name" name="name" required>
        <button type="submit">Add Equipment</button>
    </form>
</body>
</html>

**Reply by abose** [reply]
Can you attach the file with the problem here to take a look

**Reply by Olaes68** [reply]
Please see attached picture of the error in line 14.

Thanks.


![IMG_0839](https://github.com/user-attachments/assets/e4801331-a6f8-4d9c-9f95-36dda7a17760)

**Reply by abose** [reply]
@golaes The `<!DOCTYPE html>` string should come as the first line by the HTML spec. The `<?php` section should only come after the `<!DOCTYPE html>` line. That is the reason for the error.

**Reply by runningdave** [reply]
In this case the PHP Code has to start in the first line. This is quite often necessary, if PHP has to send header information. So usually all documents that contain php and html elements in one file start with <?php first to avoid such problems.

**Reply by runningdave** [reply]
I would love to see a direct preview of html code in php files, too. This is perfect to work on mixed templates. Just render the php file as html then all php code gets ignored as an unknown html-tag. Dreamweaver has had this functionality. The previews of php files diddn't look good, but it was very usefull to quickly find certain elements or texts in large template files.

**Reply by abose** [reply]
Forgot to update here, so we investigated why this is happenings, and the actual fix for this is to integrate PHP language server like Dreamweaver. But that is not scheduled for the next 3 months as it needs a lot of ground work to be in place.
